### DISCUSSION TITLE
Do all workspace: dependencies of a published package always need to be published too?

### DISCUSSION THREAD

**Discussion by jschuur** [pnpm/pnpm#6038] [Q&A]
I have a pnpm monorepo with an app that has an internal package as a `workspace:*` dependency ([sample repo](https://github.com/jschuur/pnpm-workspace-publish)):

```json
  "dependencies": {
    "@joostschuur/hello": "workspace:*"
  }
```

If I publish the app to an npm registry, do I always need to publish each of my local `workspace:` dependencies as a separate package too?

Of course, without a valid package name resolution, another package manager wouldn't know what to do with the `@joostschuur/hello` workspace dependency. I understand why this generates an error:

```
❯ npm -g i @joostschuur/jshello
npm ERR! code E404
npm ERR! 404 Not Found - GET https://registry.npmjs.org/@joostschuur%2fhello - Not found
npm ERR! 404
npm ERR! 404  '@joostschuur/hello@1.0.0' is not in this registry.
```

pnpm also [helps with versioning](https://pnpm.io/workspaces#publishing-workspace-packages) when publishing, which is good.

After I also published `@joostschuur/hello`, I could successfully do an npm install of the sample `jshello` app that used it.

My question is really whether there is some other approach I'm missing here. No way to explicitly include these workspace dependencies in my published files and then have local references to them?

Even if I bundled everything to a single file, wouldn't an npm install process still see dependencies listed and then try and install them and run into the same error?

**Comment by HuakunShen** [comment]
I have the same question, have you solved it @jschuur ?
The only solution I can think of is remove the dependency from `package.json` before publish.

**Reply by PabloStockhausen** [reply]
@HuakunShen, did you find a solution in the end? I am currently facing the same issue and would like to avoid having to publish all workspace packages.

**Reply by HuakunShen** [reply]
@PabloStockhausen I eventually put every dependency required in the package I need to publish.
Here is a possible solution but I haven't tried it.
I believe `"devDependencies"` aren't installed when a new package is added, try including your workspace package in it.
And bundle the library to JS, so the workspace package content is included.

If `"devDependencies"` doesn't work, then you can try to write a publish script, which removed all `"workspace:*"`, publish, then add it back.

Or another solution, suppose your source code is in package A, create another workspace package B. Change the bundle output path of your package A to `packageB/dist`. The `package.json` of package B won't contain `"workspace:*"`
